ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/ScriptHandlers/Paged/PlaylistScriptHandler.swift

```diff
@@ -26,32 +26,17 @@ protocol PlaylistScriptHandlerDelegate: NSObject {
   func showPlaylistOnboarding(tab: Tab?)
 }
 
-class PlaylistScriptHandler: NSObject, TabContentScript {
+class PlaylistScriptHandler: NSObject, TabContentScript, TabObserver {
   public weak var delegate: PlaylistScriptHandlerDelegate?
   private var url: URL?
-  private var urlObserver: NSObjectProtocol?
   private var asset: AVURLAsset?
   private static let queue = DispatchQueue(label: "com.playlisthelper.queue", qos: .userInitiated)
 
   init(tab: Tab) {
     self.url = tab.url
     super.init()
 
-    urlObserver = tab.webView?.observe(
-      \.url,
-      options: [.new],
-      changeHandler: { [weak self, weak tab] _, change in
-        guard let self = self, let url = change.newValue else { return }
-        if self.url != url {
-          self.url = url
-          self.asset?.cancelLoading()
-          self.asset = nil
-
-          self.delegate?.updatePlaylistURLBar(tab: tab, state: .none, item: nil)
-        }
-      }
-    )
-
+    tab.addObserver(self)
     tab.webView?.addGestureRecognizer(
```

Comment:
do we need to remove the observer from the tab when this handler is gone?

Reply:
It is removed just below when the tab is removed

Reply:
Technically its weakly held but at the moment the `AnyTabObserver` will stick around but not do anything. This is OK for now because the `*ScriptHandler` types are usually created once during the tab creation and never removed.  In the future I would introduce something in the protocol to let observers know the tab is being removed and that they should remove themselves as observer